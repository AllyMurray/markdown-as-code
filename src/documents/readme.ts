import { MarkdownDocument, MarkdownOptions } from './markdown.js';
import type { Optional } from '../@types/optional.js';
import { DocumentSection } from '../index.js';
import {
  AcknowledgementOptions,
  Acknowledgements,
  acknowledgementsSection,
} from '../sections/acknowledgements.js';
import {
  ApiReference,
  ApiReferenceOptions,
  apiReferenceSection,
} from '../sections/api-reference.js';
import {
  Appendix,
  AppendixOptions,
  appendixSection,
} from '../sections/appendix.js';
import {
  Authors,
  AuthorsOptions,
  authorsSection,
} from '../sections/authors.js';
import {
  Contributing,
  ContributingOptions,
  contributingSection,
} from '../sections/contributing.js';
import {
  EnvironmentVariables,
  EnvironmentVariablesOptions,
  environmentVariablesSection,
} from '../sections/environment-variables.js';
import {
  Examples,
  ExamplesOptions,
  examplesSection,
} from '../sections/examples.js';
import { FAQ, FAQOptions, faqSection } from '../sections/faq.js';
import {
  Installation,
  InstallationOptions,
  installationSection,
} from '../sections/installation.js';
import {
  Roadmap,
  RoadmapOptions,
  roadmapSection,
} from '../sections/roadmap.js';
import {
  RunLocalOptions,
  RunLocally,
  runLocallySection,
} from '../sections/run-locally.js';
import {
  Support,
  SupportOptions,
  supportSection,
} from '../sections/support.js';

type ReadmeOptions = Optional<MarkdownOptions, 'title'>;

const sectionOrderMap = new Map<object, number>([
  [Installation, 0],
  [EnvironmentVariables, 1],
  [RunLocally, 2],
  [Roadmap, 3],
  [ApiReference, 4],
  [Examples, 5],
  [FAQ, 6],
  [Support, 7],
  [Appendix, 8],
  [Authors, 9],
  [Acknowledgements, 10],
  [Contributing, 11],
]);

export class Readme extends MarkdownDocument {
  constructor(options: ReadmeOptions) {
    super({
      ...options,
      title: options.title ?? 'README.md',
    });
  }

  private getSortOrder(section: DocumentSection) {
    const order = sectionOrderMap.get(section.constructor);
    if (order === undefined) {
      throw new Error('Unknown section type');
    }
    return order;
  }

  private addSortedSection(section: DocumentSection) {
    return this.addSection({ section, sortOrder: this.getSortOrder(section) });
  }

  public addInstallationSection(options: InstallationOptions) {
    return this.addSortedSection(installationSection(options));
  }

  public addEnvironmentVariablesSection(options: EnvironmentVariablesOptions) {
    return this.addSortedSection(environmentVariablesSection(options));
  }

  public addRunLocallySection(options: RunLocalOptions) {
    return this.addSortedSection(runLocallySection(options));
  }

  public addRoadmapSection(options: RoadmapOptions) {
    return this.addSortedSection(roadmapSection(options));
  }

  public addApiReferenceSection(options: ApiReferenceOptions) {
    return this.addSortedSection(apiReferenceSection(options));
  }

  public addExamplesSection(options: ExamplesOptions) {
    return this.addSortedSection(examplesSection(options));
  }

  public addFaqSection(options: FAQOptions) {
    return this.addSortedSection(faqSection(options));
  }

  public addSupportSection(options: SupportOptions) {
    return this.addSortedSection(supportSection(options));
  }

  public addAppendixSection(options: AppendixOptions) {
    return this.addSortedSection(appendixSection(options));
  }

  public addAuthorsSection(options: AuthorsOptions) {
    return this.addSortedSection(authorsSection(options));
  }

  public addAcknowledgementsSection(options: AcknowledgementOptions) {
    return this.addSortedSection(acknowledgementsSection(options));
  }

  public addContributingSection(options: ContributingOptions) {
    return this.addSortedSection(contributingSection(options));
  }
}

export const createReadmeDocument = (options: ReadmeOptions) =>
  new Readme(options);
