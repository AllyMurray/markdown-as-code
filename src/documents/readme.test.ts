import { createReadmeDocument } from './readme.js';

describe('Readme', () => {
  it('should', () => {
    const readme = createReadmeDocument({
      title: 'Readme as Code',
      fileName: 'Readme.md',
    })
      .acknowledgements((section) => {
        section.add({ text: 'GitHub', url: 'https://github.com' });
      })
      .apiReference((section) => {
        section.add({
          title: 'Get all items',
          httpMethod: 'GET',
          path: '/api/items',
          parameter: {
            name: 'api_key',
            type: 'string',
            description: '**Required**. Your API key',
          },
        });
      })
      .authors((section) => {
        section.add({ githubUsername: 'AllyMurray' });
      })
      .contributing((section) => {
        section.appendContent('Some additional markdown content');
      })
      .environmentVariables((section) => {
        section.add({
          name: 'API_KEY',
          defaultValue: 'YOUR-API-KEY-HERE',
        });
      })
      .examples((section) => {
        section
          .add({
            title: 'Initialize the library',
            description:
              'When creating the Readme instance the project name and description must be supplied',
            codeblock: {
              language: 'typescript',
              code: [
                'const readme = new Readme({',
                `  name: 'Example Project',`,
                `  description: 'Define and maintain your README.md through code.',`,
                '});',
              ].join('\n'),
            },
          })
          .add({
            title: 'Add installation instructions',
            codeblock: {
              language: 'typescript',
              code: [
                'const readme = new Readme({',
                `  name: 'Example Project',`,
                `  description: 'Define and maintain your README.md through code.',`,
                '});',
                '',
                'readme.installation((step) => {',
                '  step.add({',
                `    description: 'Install using npm',`,
                `    command: 'npm i readme-as-code',`,
                '  });',
                '})',
              ].join('\n'),
            },
          });
      })
      .faq((section) => {
        section.add({
          question: 'Question 1',
          answer: 'Answer 1',
        });
      })
      .installation((step) => {
        step.add({
          description: 'Install using npm',
          command: 'npm i readme-as-code',
        });
      })
      .roadmap((section) => {
        section.add({ text: 'Item 1' });
      })
      .support((section) => {
        section.appendContent('Test content');
      });

    expect(readme.synthContent()).toMatchSnapshot();
  });
});
