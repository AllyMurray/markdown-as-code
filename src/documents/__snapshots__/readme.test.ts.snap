// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Readme > should match snapshot 1`] = `
"# Readme as Code

## Table of Contents

- [Installation](#installation)
- [Environment Variables](#environment-variables)
- [Run Locally](#run-locally)
- [Roadmap](#roadmap)
- [API Reference](#api-reference)
- [Examples](#examples)
- [FAQ](#faq)
- [Support](#support)
- [Appendix](#appendix)
- [Authors](#authors)
- [Acknowledgements](#acknowledgements)
- [Contributing](#contributing)

## Installation

Install using npm

\`\`\`bash
npm i readme-as-code
\`\`\`

## Environment Variables

To run this project, you will need to add the following environment variables to your .env file

\`\`\`bash
API_KEY=YOUR-API-KEY-HERE
\`\`\`

## Run Locally

- Run the tests

\`\`\`bash
npm t
\`\`\`

## Roadmap

- Item 1

## API Reference

#### Get all items

\`\`\`http
GET /api/items
\`\`\`

| Parameter | Type     | Description                |
| :-------- | :------- | :------------------------- |
|api_key | string | **Required**. Your API key|

## Examples

### Initialize the library

When creating the Readme instance the project name and description must be supplied

\`\`\`typescript
const readme = new Readme({
  name: 'Example Project',
  description: 'Define and maintain your README.md through code.',
});
\`\`\`

### Add installation instructions

\`\`\`typescript
const readme = new Readme({
  name: 'Example Project',
  description: 'Define and maintain your README.md through code.',
});

readme.installation((step) => {
  step.add({
    description: 'Install using npm',
    command: 'npm i readme-as-code',
  });
})
\`\`\`


## FAQ

#### Question 1

Answer 1

## Support

Test content

## Appendix

Some content

## Authors

- [@AllyMurray](https://www.github.com/AllyMurray)

## Acknowledgements

- [GitHub](https://github.com)

## Contributing

Contributions are always welcome!

See \`contributing.md\` for ways to get started.

Please adhere to this project's \`code of conduct\`.

Some additional markdown content"
`;

exports[`Readme > should write a document to disk 1`] = `
"# Test

## Table of Contents
"
`;
